lib/active_record_upsert/active_record/relation.rb

```diff
@@ -1,39 +1,36 @@
 module ActiveRecordUpsert
   module ActiveRecord
     module RelationExtensions
-      def upsert(values, wheres) # :nodoc:
-        primary_key_value = nil
-
-        if primary_key && Hash === values
-          primary_key_value = values[values.keys.find { |k|
-            k.name == primary_key
-          }]
-        end
-
+      def upsert(existing_attributes, upsert_attributes, wheres) # :nodoc:
         im = arel_table.create_insert
         im.into arel_table
 
-        substitutes, binds = substitute_values values
+        substitutes, binds = substitute_values(existing_attributes)
         column_arr = self.klass.upsert_keys || [primary_key]
         column_name = column_arr.join(',')
 
         cm = arel_table.create_on_conflict_do_update
         cm.target = arel_table[column_name]
         cm.wheres = wheres
         filter = ->(o) { [*column_arr, 'created_at'].include?(o.name) }
+        filter2 = ->(o) { upsert_attributes.include?(o.name) }
```

Comment:
blocking: i'd prefer for all filters to respond to either select or reject, not a mix.
non-blocking: if the second filter is filter2, can we rename the first to filter1?

Reply:
@humzashah this whole code is a bit crappy. I'll try to refactor it.